### PR TITLE
Helix Integration.

### PR DESCRIPTION
**Pull request by michaelgsharp**
This PR changes the CI testing process to use Helix instead of testing on the build machines (except for the code coverage tests currently). This lets us take advantage of more powerful machines for our testing. 

I disabled 6 tests tracked here #5845 due to slight differences between the results in the helix machines and the build machines.

This PR also fixes a bug in the MatrixFactorizationNative where we were compiling the native code to be compatible with Intel CPU's only.

### CODE REVIEW

**Comment by justinormont** [comment]
Description of helix: _(I had look up helix)_
> Helix is the distributed test platform that we use to run tests.  We build a helix payload that contains the publish directory of every test project that we want to test send a job with with this payload to a set of queues for the various combinations of OS that we want to test for example: `Windows.10.Amd64.ClientRS4.VS2017.Open`, `OSX.1012.Amd64.Open`, `Ubuntu.1804.Amd64.Open`. Helix takes care of unzipping, running the job, and reporting results.

via: https://github.com/dotnet/aspnetcore/blob/83f7970b8cf3d87299d564e713caa0310eaf326b/docs/Helix.md

File: test/Directory.Build.props [hunk 1]
```diff
@@ -8,7 +8,7 @@
     <RuntimeIdentifier Condition="'$(TargetFramework)' == 'net461'">win-x64</RuntimeIdentifier>
     <IsPackable>false</IsPackable>
     <GenerateDocumentationFile>true</GenerateDocumentationFile>
-
+    <GenerateRuntimeConfigurationFiles>true</GenerateRuntimeConfigurationFiles>
```
Comment (eerhardt) [inline test/Directory.Build.props:11]: Why is this necessary?

File: test/Microsoft.ML.TestFrameworkCommon/Microsoft.ML.TestFrameworkCommon.csproj [hunk 2]
```diff
@@ -8,4 +8,8 @@
     </None>
   </ItemGroup>
 
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Extensions.DependencyModel" Version="$(MicrosoftExtensionsDependencyModelVersion)" />
```
Comment (eerhardt) [inline test/Microsoft.ML.TestFrameworkCommon/Microsoft.ML.TestFrameworkCommon.csproj:12]: Why is this necessary?

File: test/Directory.Build.props (hunk 1 above)
Comment (michaelgsharp) [reply test/Directory.Build.props:11]: It was needed to create the app.config files in the publish folders. I have since switched to not need to actually run the `publish` command so I will remove it.

File: test/Microsoft.ML.TestFrameworkCommon/Microsoft.ML.TestFrameworkCommon.csproj (hunk 2 above)
Comment (michaelgsharp) [reply test/Microsoft.ML.TestFrameworkCommon/Microsoft.ML.TestFrameworkCommon.csproj:12]: It was needed in the publish step due to a version conflict with the code coverage nuget we are using. I have since switched to not need to actually run the `publish` command so I will remove it.

File: test/Microsoft.ML.Predictor.Tests/TestPredictors.cs [hunk 3]
```diff
@@ -277,7 +277,7 @@ public void BinaryClassifierLogisticRegressionTest()
             Done();
         }
 
-        [Fact]
+        [Fact(Skip = "Temporarily skipping while helix issues are resolved.")] //[Fact]
```
Comment (tarekgh) [inline test/Microsoft.ML.Predictor.Tests/TestPredictors.cs:280]: >//[Fact] [](http://example.com/codeflow?start=79&length=8)

nit. I think this not needed.

**Review by tarekgh — APPROVED** [review]
LGTM. I'll let @safern and @ericstj take a look as they more familiar with such changes :-)

File: test/Microsoft.ML.Predictor.Tests/TestPredictors.cs (hunk 3 above)
Comment (michaelgsharp) [reply test/Microsoft.ML.Predictor.Tests/TestPredictors.cs:280]: I left those in a couple of place just so I knew what was there before. But now that I say that I think they are all facts, and I can look at git history to see anyways. I'll go ahead and remove them.

Comment (eerhardt) [reply test/Microsoft.ML.Predictor.Tests/TestPredictors.cs:280]: We should be logging an issue to track re-enabling these tests. And then the `Skip` message should have the link to the issue.

Comment (michaelgsharp) [reply test/Microsoft.ML.Predictor.Tests/TestPredictors.cs]: I have an issue logged and I put it in the description of the PR, but you are right in the skip itself is much better. I'll update the message with the issue.

File: .vsts-dotnet-ci.yml [hunk 4]
```diff
@@ -67,6 +70,7 @@ jobs:
     vsTestConfiguration: "/Framework:.NETCoreApp,Version=v3.0"
     pool:
       name: Hosted VS2017
+    helixQueue: Windows.10.Amd64.Open
```
Comment (safern) [inline .vsts-dotnet-ci.yml:73]: Why not rather than duplicating these values, set the queue on `job-template` based on the pool name or other parameter. The container based runs can get complicated and it might be simpler to have all the queues at one location and have logic to decide when to use which.

See in `dotnet/runtime` is a little bit more complex but is nice to avoid duplicating queues: https://github.com/dotnet/runtime/blob/main/eng/pipelines/libraries/helix-queues-setup.yml

File: Directory.Build.targets [hunk 5]
```diff
@@ -33,7 +33,23 @@
       <NativeAssemblyReference Remove="FastTreeNative"/>
       <NativeAssemblyReference Remove="SymSgdNative"/>
       <NativeAssemblyReference Remove="MklProxyNative"/>
+      <NativeAssemblyReference Remove="libiomp5md"/>
     </ItemGroup>
+  </Target>
+
+  <Target Name="CopyNativeAssembilesPublish" AfterTargets="Publish" DependsOnTargets="SetCopyProperties">
```
Comment (safern) [inline Directory.Build.targets:40]:
```suggestion
  <Target Name="CopyNativeAssembliesPublish" AfterTargets="Publish" DependsOnTargets="SetCopyProperties">
```

File: build/ci/send-to-helix.yml [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      HelixAccessToken: ${{ parameters.HelixAccessToken }}
+      IncludeDotNetCli: ${{ parameters.IncludeDotNetCli }}
+      EnableXUnitReporter: ${{ parameters.EnableXUnitReporter }}
+      WaitForWorkItemCompletion: ${{ parameters.WaitForWorkItemCompletion }}
+      HelixBaseUri: ${{ parameters.HelixBaseUri }}
+      SYSTEM_ACCESSTOKEN: $(System.AccessToken)
+      Creator: ${{ parameters.Creator }}
+    condition: and(${{ parameters.condition }}, eq(variables['Agent.Os'], 'Windows_NT'))
+    continueOnError: ${{ parameters.continueOnError }}
+  - script: export PATH=$BUILD_SOURCESDIRECTORY/.dotnet:$PATH && $BUILD_SOURCESDIRECTORY/eng/common/msbuild.sh $BUILD_SOURCESDIRECTORY/eng/helix.proj /t:Test /bl:$BUILD_SOURCESDIRECTORY/artifacts/log/${{ parameters.Configuration }}/SendToHelix.binlog /p:Configuration=${{ parameters.Configuration }} /p:TargetArchitecture=${{ parameters.Architecture }} --warnAsError false
```
Comment (safern) [inline build/ci/send-to-helix.yml:40]: Usually when having an entry for Windows and an entry for Unix it is hard to maintain both up to date when a new parameter is added. I would suggest instead taking as a parameter the build script and then the environment variables set then under the `env:` task parameter.

For example see: https://github.com/dotnet/runtime/blob/98be10a336f216127470214fedf928ba68635c60/eng/pipelines/libraries/helix.yml#L18

In that example the command is a variable we set at the job level rather than a parameter:
https://github.com/dotnet/runtime/blob/b482c04080bb86f151e8fe3d6f2fe61649c1b5fa/eng/pipelines/common/xplat-setup.yml#L62-L63
https://github.com/dotnet/runtime/blob/b482c04080bb86f151e8fe3d6f2fe61649c1b5fa/eng/pipelines/common/xplat-setup.yml#L78-L79

File: build/ci/send-to-helix.yml [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  HelixBaseUri: 'https://helix.dot.net/' # optional -- sets the Helix API base URI (allows targeting int)
+  Creator: 'ML.NET'                      # optional -- if the build is external, use this to specify who is sending the job
+  DisplayNamePrefix: 'Run Helix Tests'   # optional -- rename the beginning of the displayName of the steps in AzDO
+  continueOnError: false                 # optional -- determines whether to continue the build if the step errors; defaults to false
+  condition: succeeded()                 # optional -- condition for step to execute; defaults to succeeded()
+
+steps:
+  - powershell: '$env:Path = "$env:BUILD_SOURCESDIRECTORY\.dotnet;$env:Path"; powershell "$env:BUILD_SOURCESDIRECTORY\eng\common\msbuild.ps1 $env:BUILD_SOURCESDIRECTORY\eng\helix.proj /t:Test /bl:$env:BUILD_SOURCESDIRECTORY\artifacts\log\${{ parameters.Configuration }}\SendToHelix.binlog /p:Configuration=${{ parameters.Configuration }} /p:TargetArchitecture=${{ parameters.Architecture }} -warnAsError 0"'
+    displayName: ${{ parameters.DisplayNamePrefix }} (Windows)
+    env:
```
Comment (safern) [inline build/ci/send-to-helix.yml:23]: When calling MSBuild using environment variables for the paramerters it makes mistakes harder to diagnose, consider using MSBuild parameters like in: https://github.com/dotnet/runtime/blob/b482c04080bb86f151e8fe3d6f2fe61649c1b5fa/eng/pipelines/libraries/helix.yml#L20-L30

File: build/ci/send-to-helix.yml [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  HelixBaseUri: 'https://helix.dot.net/' # optional -- sets the Helix API base URI (allows targeting int)
+  Creator: 'ML.NET'                      # optional -- if the build is external, use this to specify who is sending the job
+  DisplayNamePrefix: 'Run Helix Tests'   # optional -- rename the beginning of the displayName of the steps in AzDO
+  continueOnError: false                 # optional -- determines whether to continue the build if the step errors; defaults to false
+  condition: succeeded()                 # optional -- condition for step to execute; defaults to succeeded()
+
+steps:
+  - powershell: '$env:Path = "$env:BUILD_SOURCESDIRECTORY\.dotnet;$env:Path"; powershell "$env:BUILD_SOURCESDIRECTORY\eng\common\msbuild.ps1 $env:BUILD_SOURCESDIRECTORY\eng\helix.proj /t:Test /bl:$env:BUILD_SOURCESDIRECTORY\artifacts\log\${{ parameters.Configuration }}\SendToHelix.binlog /p:Configuration=${{ parameters.Configuration }} /p:TargetArchitecture=${{ parameters.Architecture }} -warnAsError 0"'
```
Comment (safern) [inline build/ci/send-to-helix.yml:21]: .dotnet should already be added to the path when it is acquired by arcade.

Comment (safern) [reply build/ci/send-to-helix.yml:21]: Also, you don't need to do `$env:BUILD_SOURCESDIRECTORY` you can access this as a yaml variable and the yaml engine will translate them to the correct value. i.e `$(Build.SourcesDirectory)`

File: eng/helix.proj [hunk 9]
```diff
@@ -0,0 +1,152 @@
+<Project Sdk="Microsoft.DotNet.Helix.Sdk" InitialTargets="CreateHelixWorkItems">
+	<Choose>
+		<When Condition="$(HelixTargetQueues.ToLowerInvariant().Contains('windows'))">
```
Comment (safern) [inline eng/helix.proj:3]: We usually use spaces and 2 spaces for indentation on MSBuild projects.

File: build/ci/send-to-helix.yml (hunk 6 above)
Comment (michaelgsharp) [reply build/ci/send-to-helix.yml:40]: I did it this way because its the way the `send-to-helix.yml` file from Arcade is done and I just followed their example. For the actual build command we do it the way you suggest though. If you feel strongly that I should change it I will, but I think leaving it as close to their example as possible will be easier for others who aren't as familiar with the process to understand whats going on.

File: build/ci/send-to-helix.yml (hunk 7 above)
Comment (michaelgsharp) [reply build/ci/send-to-helix.yml:23]: That line was taken directly from the `send-to-helix.yml` from Arcade. I don't know why they chose to do it that way, but my guess is that it makes it easier to try and debug locally. You can't run the yml file locally, but you can set up the env variable the same way so you can use the script the same way. I think again if you feel strongly its better the other way I'm happy to switch it, but otherwise I think I would rather keep it this way to be consistent with what Arcade has.

File: build/ci/send-to-helix.yml (hunk 8 above)
Comment (michaelgsharp) [reply build/ci/send-to-helix.yml:21]: Good to know. I was having some issues and added that just to be safe. I don't think it did anything but I left it because I coudln't confirm. But since you know I will go ahead and remove it.

File: eng/helix.proj (hunk 9 above)
Comment (michaelgsharp) [reply eng/helix.proj:3]: Sounds good. I will update this file with that.

File: .vsts-dotnet-ci.yml (hunk 4 above)
Comment (michaelgsharp) [reply .vsts-dotnet-ci.yml:73]: Do you think all the extra complexity with the conditionals will be beneficial in our case? We don't have nearly as many queues or situations as the runtime does. We only set one queue per job, does that change anything?

If you think we should still do it lets sync tomorrow on this. I'm not super familiar with the yml format for AzDo and it will be much faster if we do it together.

Comment (safern) [reply .vsts-dotnet-ci.yml:73]: Maybe for your scenario the extra complexity is not worth it. If you start adding more and more queues, maybe consider revisiting.

File: build/ci/send-to-helix.yml (hunk 6 above)
Comment (safern) [reply build/ci/send-to-helix.yml]: The reason why the arcade `send-to-helix.yml` is like that, is because it is a more generic file for repos to share and use, so it would be harder user experience to use a predefined variable because for templates you usually just read parameters. I did propose that template to take a parameter for the msbuild script command, however they wanted to keep it simple and repos to not have to think about checking for windows vs unix. 

However, since this is your custom send to helix, I would definitely encourage a cleaner/smaller approach rather than duplicating, it helps with readability and maintainability in the future when you add more properties that need to be set for the helix proj.

File: build/ci/send-to-helix.yml (hunk 7 above)
Comment (safern) [reply build/ci/send-to-helix.yml:23]: I think it is easier to debug locally to copy a command rather than having to set multiple environment variables manually. Also when looking at binlogs we sometimes forget that environment variables come into play in MSBuild, so they are "obscure" for some people and harder to debug as the first thing and maybe only thing we look on a binlog are properties.

File: build/ci/send-to-helix.yml (hunk 6 above)
Comment (michaelgsharp) [reply build/ci/send-to-helix.yml]: That makes sense. I'll go ahead and switch it.

File: .vsts-dotnet-ci.yml (hunk 4 above)
Comment (michaelgsharp) [reply .vsts-dotnet-ci.yml:73]: I will be adding a few more. For arm/arm64/m1. For those though we will only have a debug/release build for each. Is that enough you think to make it better to add in this logic?

Comment (safern) [reply .vsts-dotnet-ci.yml:73]: I don't have a strong opinion. If you ever feel like it is getting too much repetition might be worth exploring the other option.

File: eng/helix.proj [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+    <!-- The other DotNetCliRuntimes are figured out correctly by the queue name -->
+    <DotNetCliRuntime Condition="'$(BuildArchitecture)' == 'x86'">win-x86</DotNetCliRuntime>
+
+    <!-- 'true' to produce a build error when tests fail. Default 'true' -->
+    <FailOnTestFailure>true</FailOnTestFailure>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <AdditionalDotNetPackage Include="$(MicrosoftNETCore3PlatformsVersion)">
+    <PackageType>runtime</PackageType>
```
Comment (safern) [inline eng/helix.proj:41]: Nit: indentation missing here.

File: eng/helix.proj [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+    <DotNetCliRuntime Condition="'$(BuildArchitecture)' == 'x86'">win-x86</DotNetCliRuntime>
+    </AdditionalDotNetPackage>
+
+    <AdditionalDotNetPackage Include="$(MicrosoftNETCorePlatformsVersion)">
+    <PackageType>runtime</PackageType>
```
Comment (safern) [inline eng/helix.proj:46]: Nit, indentation missing.

File: eng/helix.proj [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+    <DotNetCliRuntime Condition="'$(BuildArchitecture)' == 'x86'">win-x86</DotNetCliRuntime>
+    </AdditionalDotNetPackage>
+  </ItemGroup>
+
+  <PropertyGroup Condition="'$(HelixType)' == ''">
+    <!-- For PRs we want helixtype to be the same for all frameworks except package testing-->
+    <TestScope Condition="'$(TestScope)' == ''">innerloop</TestScope>
+    <HelixType>test/unit/cli/$(TestScope)/</HelixType>
+   </PropertyGroup>
```
Comment (safern) [inline eng/helix.proj:55]: Indentation is off.

File: eng/helix.proj [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+    <Delete Condition="$(HelixTargetQueues.ToLowerInvariant().Contains('osx'))"
+      Files="@(LinuxFiles);@(WindowsFiles)" />
+
+    <Delete Condition="$(HelixTargetQueues.ToLowerInvariant().Contains('ubuntu'))"
+      Files="@(WindowsFiles);@(OsxFiles)" />
+
+    <ItemGroup>
+      <HelixWorkItem Include="%(ProjectsWithTargetFramework.Filename)">
+        <PayloadDirectory>$(BUILD_SOURCESDIRECTORY)\artifacts\bin\%(ProjectsWithTargetFramework.Filename)\$(BuildConfig)\%(ProjectsWithTargetFramework.TargetFrameworks)\$(PublishFolder)</PayloadDirectory>
+        <Command Condition="$(IsPosixShell)">dotnet exec --roll-forward Major --runtimeconfig %(ProjectsWithTargetFramework.Filename).runtimeconfig.json --depsfile %(ProjectsWithTargetFramework.Filename).deps.json $HELIX_CORRELATION_PAYLOAD/xunit-runner/tools/netcoreapp2.0/xunit.console.dll %(ProjectsWithTargetFramework.Filename).dll -notrait Category=SkipInCI -xml testResults.xml</Command>
```
Comment (safern) [inline eng/helix.proj:144]: NIT: 
You could avoid duplicating the command and just changing the env var access syntax if you capture the syntax in a property. i.e:

`<HelixCorrelationPayload Condition=IsPosix>%HELIX%</HelixCorrelationPayload>`

And then you can use that property.

**Review by safern — APPROVED** [review]

File: eng/helix.proj (hunk 13 above)
Comment (michaelgsharp) [reply eng/helix.proj:144]: So `%HELIX%` maps to the correct path then? Let me try that and see if it works. I'll fix the spacing too.

Comment (safern) [reply eng/helix.proj:144]: No, %HELIX% doesn't map to the correct path, that was just an example of how you could leverage the decision of the syntax you want to use with a property.

File: eng/helix.proj (hunk 10 above)
Comment (michaelgsharp) [reply eng/helix.proj:41]: Fixed.

File: eng/helix.proj (hunk 11 above)
Comment (michaelgsharp) [reply eng/helix.proj:46]: Fixed.

File: eng/helix.proj (hunk 12 above)
Comment (michaelgsharp) [reply eng/helix.proj:55]: Fixed.

File: eng/helix.proj (hunk 13 above)
Comment (michaelgsharp) [reply eng/helix.proj]: Yeah, I figured that much when I took a second look. lol. But I got it correct now and only have the one command now.

I'm gonna go ahead and merge this in.
